### PR TITLE
feat: Add UI for Snaps pre-approved connections

### PR DESCRIPTION
**Pull request by david0xd**
## **Description**
This PR enables pre-approved connections to be displayed in the UI of the permission confirmation step when installing or updating a Snap.

#### Notes
- Given that pre-approved connections are defined in a specific way, different from the permission specification and need to be displayed as a part of permission list (permission confirmation process), they're integrated in the permission display process as a side effect.
- Pre-approved connection is considered as a transient permission that is handled by couple of side effect functions and components in order to be displayed alongside other permissions and treated equaly.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24403?quickstart=1)

## **Related issues**
Fixes: https://github.com/MetaMask/MetaMask-planning/issues/1895

## **Manual testing steps**
1. Create a Snap with `initialConnections` specified. For example:
```json
"initialConnections": {
    "https://snaps.metamask.io/": {},
    "https://app.uniswap.org/": {},
    "https://metamask.github.io/": {},
    "npm:@web3-antivirus/web3-antivirus-snap": {},
    "npm:@near-snap/plugin": {}
  }
```
2. Try installing a Snap.
3. Verify that connections are appearing in the UI alongside permissions.
4. Try the same process for the Snap update flow.

### **Before**
No changes to the UI implemented before. Pre-approved connections are just added to the permission list.

### **After**
Screenshot 2024-05-20 at 11 56 34
Screenshot 2024-05-20 at 11 56 51
Screenshot 2024-05-23 at 17 44 46
Screenshot 2024-05-23 at 17 45 04

## **Pre-merge author checklist**
- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**
- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/pages/permissions-connect/snaps/snap-update/snap-update.js [hunk 1]
```diff
@@ -186,7 +189,11 @@ export default function SnapUpdate({
                 approvedPermissions={approvedPermissions}
                 revokedPermissions={revokedPermissions}
                 newPermissions={newPermissions}
+                approvedConnections={approvedConnections}
+                revokedConnections={revokedConnections}
+                newConnections={newConnections}
                 targetSubjectMetadata={targetSubjectMetadata}
+                connections={requestState.connections || {}}
```
Comment (FrederikBolding) [inline ui/pages/permissions-connect/snaps/snap-update/snap-update.js:196]: This doesn't exist anymore

File: ui/helpers/utils/permission.js [hunk 2]
```diff
@@ -552,7 +554,48 @@ export const PERMISSION_DESCRIPTIONS = deepFreeze({
 
     return result;
   },
+  [ConnectionPermission.connection_permission]: ({
```
Comment (FrederikBolding) [inline ui/helpers/utils/permission.js:557]: Please document that this is a pseudo-permission for display purposes

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+} from '../../../../helpers/constants/design-system';
+import { getAvatarFallbackLetter } from '../../../../helpers/utils/util';
+import PermissionCell from '../../permission-cell';
+import { useI18nContext } from '../../../../hooks/useI18nContext';
+
+export default function SnapPermissionCellDisplay({
+  snapId,
+  showOptions,
+  permission,
+  subjectMetadata,
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js:29]: We should document that this is the subject metadata of the target of the permission. Maybe even rename 🤔

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.test.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
+import { screen } from '@testing-library/react';
+import React from 'react';
+import { renderWithProvider } from '../../../../../test/jest';
+import SnapPermissionCellDisplay from './snap-permission-cell-display';
+
+describe('Snap Permission List', () => {
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.test.js:6]: This seems out of date

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  FontWeight,
+  JustifyContent,
+  TextColor,
+  TextVariant,
+} from '../../../../helpers/constants/design-system';
+import { getAvatarFallbackLetter } from '../../../../helpers/utils/util';
+import PermissionCell from '../../permission-cell';
+import { useI18nContext } from '../../../../hooks/useI18nContext';
+
+export default function SnapPermissionCellDisplay({
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js:25]:
```suggestion
export default function SnapPermissionCell({
```
WDYT?

File: shared/constants/permissions.ts [hunk 6]
```diff
@@ -21,6 +21,12 @@ export const RestrictedMethods = Object.freeze({
   ///: END:ONLY_INCLUDE_IF
 } as const);
 
+///: BEGIN:ONLY_INCLUDE_IF(snaps)
+export const ConnectionPermission = Object.freeze({
```
Comment (FrederikBolding) [inline shared/constants/permissions.ts:25]: Document that this isn't really a permission

File: ui/components/app/snaps/update-snap-permission-list/update-snap-permission-list.js [hunk 7]
```diff
@@ -91,5 +114,21 @@ UpdateSnapPermissionList.propTypes = {
...
+  revokedConnections: PropTypes.object.isRequired,
+  /**
+   * New pre-approved connections that are being requested
+   */
+  newConnections: PropTypes.object.isRequired,
   targetSubjectMetadata: PropTypes.object.isRequired,
+  /**
+   * Pre-approved connections used by Snaps for Snaps or websites.
+   */
+  connections: PropTypes.object,
```
Comment (FrederikBolding) [inline ui/components/app/snaps/update-snap-permission-list/update-snap-permission-list.js:133]: This doesn't exist anymore

Comment (david0xd) [reply ui/components/app/snaps/update-snap-permission-list/update-snap-permission-list.js]: Removed.

File: shared/constants/permissions.ts (hunk 6 above)
Comment (david0xd) [reply shared/constants/permissions.ts:27]: Added comment.

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js (hunk 5 above)
Comment (david0xd) [reply ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js]: Renamed.

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js (hunk 3 above)
Comment (david0xd) [reply ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell-display.js]: Renamed, should be more clear now.

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell.js
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell.js:1]: Rename folder here?

File: ui/selectors/selectors.js [hunk 8]
```diff
@@ -1157,6 +1157,34 @@ export const getMemoizedInterfaceContent = createDeepEqualSelector(
...
+const selectOrigins = (_state, origins) => origins;
+
+/**
+ * Retrieve metadata for multiple subjects (origins).
+ *
+ * @param state - Redux state object.
+ * @param origins - Object containing keys that represent subject's identification.
+ * @returns Key:value object containing metadata attached to each subject key.
+ */
+export const getMultipleTargetsSubjectMetadata = createSelector(
```
Comment (FrederikBolding) [inline ui/selectors/selectors.js:1177]: Can this be a deep equal memoized selector?

File: ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell.js
Comment (david0xd) [reply ui/components/app/snaps/snap-permission-cell-display/snap-permission-cell.js:1]: Renamed.

File: ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+  if (permission.connection) {
+    if (connectionSubjectMetadata?.subjectType === SubjectType.Snap) {
+      label = t('snapConnectTo', [
+        <Text
+          key="connectToMain"
+          fontWeight={FontWeight.Medium}
+          variant={TextVariant.inherit}
+          color={TextColor.inherit}
+        >
+          {connectionSubjectMetadata.name}
```
Comment (FrederikBolding) [inline ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js:46]: We can't really use `connectionSubjectMetadata.name` for Snaps as they may have a localized name. 

That's what we use `getSubjectName` for in `permission.js`

Comment (david0xd) [reply ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js]: Updated.

**Review by FrederikBolding — DISMISSED** [review]

**Review by FrederikBolding — DISMISSED** [review]

File: ui/components/app/permission-cell/permission-cell-status.js [hunk 10]
```diff
@@ -111,6 +118,18 @@ export const PermissionCellStatus = ({ revoked, dateApproved, accounts }) => {
...
+
+  if (revoked) {
+    statusToShow = statusRevoked;
+  } else if (dateApproved) {
+    statusToShow = statusApprovedOn;
+  } else if (approved) {
+    statusToShow = statusApproved;
+  } else {
+    statusToShow = statusRequestedNow;
+  }
```
Comment (Mrtenz) [inline ui/components/app/permission-cell/permission-cell-status.js:131]: It's cleaner to extract this and the stuff above into a function and return the message.

```ts
function getStatusMessage() {
  if (revoked) {
    return accounts && accounts.length
      ? t('permissionRevokedForAccounts', [renderAccountsGroup()]);
  }

  if (dateApproved) {
    return accounts && accounts.length
      ? t('approvedOnForAccounts', [
          formatDate(dateApproved, 'yyyy-MM-dd'),
          renderAccountsGroup(),
        ])
      : t('approvedOn', [formatDate(dateApproved, 'yyyy-MM-dd')]);
  }

  // and so on
}

File: ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+}) {
+  const t = useI18nContext();
+
+  let { label, description, leftIcon } = permission;
+
+  const snapsMetadata = useSelector(getSnapsMetadata);
+
+  if (permission.connection) {
+    if (connectionSubjectMetadata?.subjectType === SubjectType.Snap) {
+      const snapName = getSnapName(snapsMetadata)(
```
Comment (Mrtenz) [inline ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js:47]: Again cleaner to extract some stuff into functions.

File: ui/components/app/permission-cell/permission-cell-status.js (hunk 10 above)
Comment (FrederikBolding) [reply ui/components/app/permission-cell/permission-cell-status.js:131]: That pattern is a bit hard to achieve when using a hook though 🤔

Comment (david0xd) [reply ui/components/app/permission-cell/permission-cell-status.js]: Refactored / updated.

File: ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js (hunk 11 above)
Comment (david0xd) [reply ui/components/app/snaps/snap-permission-cell/snap-permission-cell.js]: Refactored / updated.

**Review by FrederikBolding — APPROVED** [review]

**Review by Mrtenz — APPROVED** [review]

**Pull request merged**